source/base/mpi.cc

```diff
@@ -812,7 +814,890 @@ namespace Utilities
+          AssertThrowMPI(ierr3);
+        }
+#endif
+    }
+
+    template <typename T1, typename T2>
+    void
+    ConsensusAlgorithm_NBX<T1, T2>::start_communication()
+    {
+#ifdef DEAL_II_WITH_MPI
+      // 1)
+      targets              = this->process.compute_targets();
+      const auto n_targets = targets.size();
+
+      // 2) allocate memory
+      recv_buffers.resize(n_targets);
+      recv_requests.resize(n_targets);
+      send_requests.resize(n_targets);
+      send_buffers.resize(n_targets);
+
+      {
+        // 4) send and receive
+        for (unsigned int i = 0; i < n_targets; i++)
+          {
+            const unsigned int rank  = targets[i];
+            const unsigned int index = i;
+
+            // translate index set to a list of pairs
+            auto &send_buffer = send_buffers[index];
+            this->process.pack_recv_buffer(rank, send_buffer);
+
+            // start to send data
+            auto ierr1 = MPI_Isend(send_buffer.data(),
+                                   send_buffer.size(),
+                                   internal::mpi_type_id(send_buffer.data()),
+                                   rank,
+                                   tag_request,
+                                   this->comm,
+                                   &send_requests[index]);
+            AssertThrowMPI(ierr1);
+
+            // start to receive data
+            auto &recv_buffer = recv_buffers[index];
+            this->process.prepare_recv_buffer(rank, recv_buffer);
+            auto ierr2 = MPI_Irecv(recv_buffer.data(),
+                                   recv_buffer.size(),
+                                   internal::mpi_type_id(recv_buffer.data()),
+                                   rank,
+                                   tag_delivery,
+                                   this->comm,
+                                   &recv_requests[index]);
+            AssertThrowMPI(ierr2);
+          }
+      }
+#endif
+    }
+
+    template <typename T1, typename T2>
+    void
+    ConsensusAlgorithm_NBX<T1, T2>::clean_up_and_end_communication()
+    {
+#ifdef DEAL_II_WITH_MPI
+      // clean up
+      {
+        auto ierr1 = MPI_Waitall(send_requests.size(),
+                                 send_requests.data(),
+                                 MPI_STATUSES_IGNORE);
+        AssertThrowMPI(ierr1);
+
+        auto ierr2 = MPI_Waitall(recv_requests.size(),
+                                 recv_requests.data(),
+                                 MPI_STATUSES_IGNORE);
+        AssertThrowMPI(ierr2);
+
+        auto ierr3 = MPI_Wait(&barrier_request, MPI_STATUS_IGNORE);
+        AssertThrowMPI(ierr3);
+
+        for (auto &i : request_requests)
+          {
+            auto ierr = MPI_Wait(i.get(), MPI_STATUS_IGNORE);
+            AssertThrowMPI(ierr);
+          }
+
+#  ifdef DEBUG
+        // note: IBarrier seems to make problem during testing, this additional
+        // Barrier seems to help
```

Comment:
Are you sure that it is not needed in `Release` mode as well then?

Reply:
As mentioned in an earlier message, we only experienced some issues during running the test suite (where multiple tests are run in parallel). The reason for this is with high probability the usage of IBarrier. However, I have never experienced problems when I was running the tests individually and we did not have any problems during full-machine runs on SuperMUC.